src/libertem/io/dataset/dm.py

```diff
@@ -0,0 +1,272 @@
+        idx = 0
+    else:
+        idx = 1
+    offset = fh.dataOffset[idx]
+    return offset
+
+
+class StackedDMFile(File3D):
+    """
+    A single file from a stack of dm files; this class reads directly
+    using offset and size, bypassing the reading of tag structures etc.
+    """
+    def __init__(self, path, start_idx, offset, shape, dtype):
+        """
+        Parameters
+        ----------
+
+        path : string
+
+        start_idx : int
+
+        offset : int
+
+        shape : (int, int)
+
+        dtype : numpy dtype
+        """
+        super().__init__()
+        self._path = path
+        self._start_idx = start_idx
+        self._offset = offset
+        self._shape = shape
+        self._size = int(np.product(shape))
+        self._dtype = np.dtype(dtype)
+        self._fh = None
+
+    @property
+    def dtype(self):
+        return self._dtype
+
+    @property
+    def shape(self):
+        return self._shape
+
+    def open(self):
+        self._fh = open(self._path, "rb")
+
+    def close(self):
+        self._fh.close()
+        self._fh = None
+
+    @property
+    def num_frames(self):
+        """
+        the stacked format contains one frame per dm file
+        """
+        return 1
+
+    @property
+    def start_idx(self):
+        return self._start_idx
+
+    def _read_frame(self):
+        # FIXME: reusing the buffer only works if there are multiple frames per file.
+        # readsize = self._size * self._dtype.itemsize
+        # buf = self.get_buffer("readbuf", readsize)
+        # FIXME: can we get rid of this buffer? maybe only if dtype matches?
+        buf = np.zeros(self._size, dtype=self._dtype)
+        self._fh.seek(self._offset)
+        size_read = self._fh.readinto(buf)
+        assert size_read == self._size * self._dtype.itemsize
+        return buf
+
+    def readinto(self, start, stop, out, crop_to=None):
+        """
+        Read [`start`, `stop`) images from this file into `out`
+        """
+        assert stop - start == 1
+        data = self._read_frame()
+
+        data = data.reshape(self.shape)
+
+        if crop_to is not None:
+            # TODO: maybe limit I/O to the cropped region, too?
+            data = data[crop_to.get(sig_only=True)]
+        out[...] = data
+
+
+class DMFileSet(FileSet3D):
+    pass
+
+
+class DMDataSet(DataSet):
+    """
+    Reader for stacks of DM3/DM4 files. Each file should contain a single frame.
+
+    Note
+    ----
+    This DataSet is not supported in the GUI yet, as the file dialog needs to be
+    updated to properly handle opening series.
+
+    Note
+    ----
+    Single-file 4D DM files are not yet supported. The use-case would be
+    to read DM4 files from the conversion of K2 data, but those data sets
+    are actually transposed (nav/sig are swapped), so it is very slow to
+    read single diffraction patterns / frames. Please let us know if you have
+    a use-case for single-file 4D DM files and we may add support!
```

Comment:
That means the data would have to be transposed back into the usual shape, which is slow, or algorithms would have to be adapted to work directly on transposed data. As an example, applying a mask in the conventional layout corresponds to calculating a weighted sum frame along the navigation dimension in the transposed layout.

Since the transposed layout corresponds to a TEM tilt series, support for transposed 4D STEM data could have more general applications beyond supporting 4D DM4 files. Please contact us if you have a use-case for single-file 4D DM files or other applications that process stacks of TEM files, and we may add support!